src/promptflow/promptflow/exceptions.py

```diff
@@ -209,3 +219,101 @@ class ValidationException(UserErrorException):
     """Exception raised when validation fails."""
 
     pass
+
+
+class _ErrorInfo:
+    _exception_codes = dict()
+
+    @classmethod
+    def get_error_info(cls, e: Exception):
+        if not isinstance(e, Exception):
+            return None, None, None, None
+
+        e = cls.select_exception(e)
+        if cls._is_system_error(e):
+            return ErrorCategory.SystemError, cls._error_type(e), cls._error_target(e), cls._error_message(e)
+
+        return ErrorCategory.UserError, cls._error_type(e), cls._error_target(e), cls._error_message(e)
+
+    @classmethod
+    def select_exception(cls, e: Exception):
+        """Select the exception  in e and e.__cause__, and prioritize the Exception defined in the SDK."""
+
+        if e.__cause__ and isinstance(e.__cause__, PromptflowException):
+            return e.__cause__
+
+        if isinstance(e, PromptflowException):
+            return e
+
+        if e.__cause__ and isinstance(e.__cause__, HttpResponseError):
+            return e.__cause__
+
+        return e
+
+    @classmethod
+    def _is_system_error(cls, e: Exception):
+        if isinstance(e, (SystemErrorException, HttpResponseError)):
+            return True
+        if hasattr(e, "status_code") or (hasattr(e, "response") and hasattr(e.response, "status_code")):
+            status_code = str(e.status_code) if hasattr(e, "status_code") else str(e.response.status_code)
+            # Except for 429, 400-499 are all client errors.
+            if not status_code.startswith("4") and status_code != "429":
+                return True
+
+        return False
+
+    @classmethod
+    def _error_type(cls, e: Exception):
+        return type(e).__name__
+
+    @classmethod
+    def _error_target(cls, e: Exception):
+        return getattr(e, "target", ErrorTarget.UNKNOWN)
+
+    @classmethod
+    def _error_message(cls, e: Exception):
+        exception_codes = cls._get_exception_codes(e)
+        msg = getattr(e, "message_format", "")
+        exception_code = exception_codes[-1]
+        for item in exception_codes[::-1]:  # Prioritize recording the location of promptflow package errors
```

Comment:
shall we do this just for promptflow system error?

Reply:
No, it includes any errors. Recording this is to know which line of code the user's error was thrown on in **promptflow**, not to record the line where the error was thrown in **site-packages**, making it easier for us to investigate where the error occurred.
![image](https://github.com/microsoft/promptflow/assets/23182548/fe528839-6ddd-4dec-85e1-1bd71669598d)

Reply:
I have modified the logic to only record exceptions when they are thrown in the promptflow package, and only the promptflow module and code will be recorded.